frontend/deb/pkg.go

```diff
@@ -0,0 +1,154 @@
+// the caller of the patches applied and is generally just more inline with
+// a typical deb build.
+//
+// This is using git instead of raw `diff` or other standalone tooling because only git appears to preserve permissions for new files.
+// As an example, if patch adds a new file with mode +x, `diff` will not see the permissions for that new file.
+func createPatches(spec *dalec.Spec, sources map[string]llb.State, worker llb.State, dr llb.State, opts ...llb.ConstraintsOpt) llb.State {
+	patches := llb.Scratch()
+	if len(spec.Patches) > 0 {
+		patchesMountInput := llb.Scratch().
+			File(llb.Mkfile("dalec-changes.patch", 0o600, patchHeader))
+
+		patches = worker.
+			Run(dalec.ShArgs("set -e; git config --global user.email phony; git config --global user.name Dalec")).
+			Run(
+				dalec.ShArgs("set -e; git init .; git add .; git commit -m 'Initial commit'; \"${DEBIAN_DIR}/dalec/patch.sh\"; git add .; git commit -m 'With patch'; git diff HEAD~1 >> /work/out/dalec-changes.patch; echo 'dalec-changes.patch' > /work/out/series"),
+				llb.Dir("/work/sources"),
+				mountSources(sources, "/work/sources", nil),
+				// DEBIAN_DIR is used by the patch script to find the debian directory where we actually have the patches
+				llb.AddEnv("DEBIAN_DIR", "/work/debian"),
+				llb.AddMount("/work/debian", dr, llb.SourcePath("debian"), llb.Readonly),
+				dalec.WithConstraints(opts...),
+			).AddMount("/work/out", patchesMountInput)
+	}
+
+	return patches
+}
+
+func SourcePackage(sOpt dalec.SourceOpts, worker llb.State, spec *dalec.Spec, targetKey, distroVersionID string, opts ...llb.ConstraintsOpt) (llb.State, error) {
+	if err := validateSpec(spec); err != nil {
+		return llb.Scratch(), err
+	}
+	dr, err := Debroot(sOpt, spec, worker, llb.Scratch(), targetKey, "", distroVersionID)
+	if err != nil {
+		return llb.Scratch(), err
+	}
+
+	sources, err := dalec.Sources(spec, sOpt)
+	if err != nil {
+		return llb.Scratch(), err
+	}
+
+	gomodSt, err := spec.GomodDeps(sOpt, worker, opts...)
+	if err != nil {
+		return llb.Scratch(), errors.Wrap(err, "error preparing gomod deps")
+	}
+
+	if gomodSt != nil {
+		sources[gomodsName] = *gomodSt
+	}
+
+	patches := createPatches(spec, sources, worker, dr, opts...)
+	debSources := debSources(worker, sources)
+
+	work := worker.Run(
+		dalec.ShArgs("set -e; ls -lh; dpkg-buildpackage -S -us -uc; mkdir -p /tmp/out; ls -lh; cp -r /work/"+spec.Name+"_"+spec.Version+"* /tmp/out; ls -lh /tmp/out"),
+		llb.Dir("/work/pkg"),
+		llb.AddMount("/work/pkg/debian", dr, llb.SourcePath("debian")), // This cannot be readonly because the debian directory gets modified by dpkg-buildpackage
+		llb.AddMount("/work/pkg/debian/patches", patches, llb.Readonly),
+		llb.AddEnv("DH_VERBOSE", "1"),
+		llb.Network(pb.NetMode_NONE),
+		mountSources(sources, "/work/pkg", sanitizeSourceKey),
+		dalec.RunOptFunc(func(ei *llb.ExecInfo) {
+			for key, src := range debSources {
```

Comment:
Can you add a comment clarifying the need for this RunOptFunc?

Reply:
Done